### PR TITLE
Adding implicit dataset ID support for GAE and GCE.

### LINKED ISSUE #475: Add GAE/GCE support for datastore.set_default_dataset and add exception on failure
See discussion in #444 for comments on exception.

Something like [`oauth2client.client._get_environment`][1] could be good to detect GAE/GCE dataset ID's but `_detect_gce_environment` should be better as discussed in google/oauth2client#93

Not blocking for M1.

[1]: https://github.com/google/oauth2client/blob/481b4b6310e5872fbd5ed6c7aa336938e564d729/oauth2client/client.py#L953

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #475

### CODE REVIEW

File: gcloud/datastore/_implicit_environ.py [hunk 1]
```diff
@@ -4,9 +4,63 @@
...
 
 DATASET_ID = None
 """Module global to allow persistent implied dataset ID from enviroment."""
 
 CONNECTION = None
 """Module global to allow persistent implied connection from enviroment."""
+
+
+def app_engine_id():
+    """Gets the App Engine application ID if it can be implied.
```
Comment (tseaver) [inline gcloud/datastore/_implicit_environ.py]: "inferred"

File: gcloud/datastore/_implicit_environ.py [hunk 2]
```diff
@@ -4,9 +4,63 @@
...
+              else ``None``.
+    """
+    if app_identity is None:
+        return None
+
+    return app_identity.get_application_id()
+
+
+def compute_engine_id():
+    """Gets the Compute Engine project ID if it can be implied.
```
Comment (tseaver) [inline gcloud/datastore/_implicit_environ.py]: "inferred" again.

File: gcloud/datastore/_implicit_environ.py [hunk 3]
```diff
@@ -4,9 +4,63 @@
...
+    """Gets the App Engine application ID if it can be implied.
+
+    :rtype: string or ``NoneType``
+    :returns: App Engine application ID if running in App Engine,
+              else ``None``.
+    """
+    if app_identity is None:
+        return None
+
+    return app_identity.get_application_id()
```
Comment (tseaver) [inline gcloud/datastore/_implicit_environ.py]: Could be:

```python
    return app_identity and app_identity.get_application_id()
```

Comment (dhermes) [inline gcloud/datastore/_implicit_environ.py]: I've been trying to make the `None` checks explicit after a previous conversation we had. Do you recall? Shall I still update this chunk of code?

Comment (tseaver) [inline gcloud/datastore/_implicit_environ.py]: I think just leave it be.

**Comment by dhermes** [comment]
@tseaver I addressed the vocab changes (incorporated them into the original commits) and rebased on top of current master.

LGTY?

**Comment by tseaver** [comment]
LGTM

**Pull request merged**